servers/jax-rs-tests/src/main/java/org/projectnessie/jaxrs/tests/BaseTestNessieApi.java

```diff
@@ -373,10 +374,36 @@ public void references() throws Exception {
       api().deleteTag().tag(tag).delete();
     }
 
+    // We cannot delete a branch if the expected hash is not reachable from its HEAD. Here,
+    // the expected hash is not reachable anymore, because the branch was reassigned to main
+    // previously.
+    // In such cases we expect a not-found error.
     AbstractThrowableAssert<?, ? extends Throwable> deleteConflict =
         isV2()
             ? soft.assertThatThrownBy(() -> api().deleteBranch().branch(branch).getAndDelete())
             : soft.assertThatThrownBy(() -> api().deleteBranch().branch(branch).delete());
+    deleteConflict
```

Comment:
This is the only test that failed after the changes. It is testing an edge case: the branch is at c1, but the provided expected hash is c2.

Previously there was no hash resolution, so the endpoint would throw a CONFLICT error since c2 is not the expected hash.

With the new hash resolution in place, the endpoint now throws NOT_FOUND because c2 is not reachable from c1 at all.

I added another test case below that generates a CONFLICT as before.

Reply:
I'm not sure about the new behavour here. In this case, the hash is valid, but it is not in the current commit chain of the reference. This is exactly the case when one client assigns the reference, while another client attempts to delete it using the old HEAD. I tend to think that a "conflict" is more appropriate from the high-level perspective.

We seems to have two general approaches. Read operations throw "not found" when the hash is not found on the branch. Write operations throw "conflict" when the "expected" hash does not match server-side expectations. The relative hash case kind of mixes both cases.

Since write operations now require a concrete base hash, could we make the resolution more lenient and defer the (final) exception throwing to the highest-level method that knows the operation's end-user semantics? 

On the other hand, all non-trivial unambiguous relative hash references resolve to something other than HEAD. Perhaps, we could simplify the parsing of the expected hash parameter for assign/deleteReference requests and respond with a "bad request" if a relative spec is used. If the hash is absolute we could avoid checking it against the history, but do "conflict" by comparing it to current HEAD right away (also requires less backend work).

Reply:
Your comments are 100% spot on. I hesitated a long time myself before picking one solution.

To summarize, we'd have 3 ways to move forward with this issue:

1. Accept the behavior change: if c2 is not in main, then `main@c2` should legitimately throw not-found.
2. Use `DETACHED` to resolve relative hashes.
  a.  Using `DETACHED`, `VersionStore.hashOnReference()` will simply lookup the commit but won't validate that it belongs to any branch.
3. respond with a "bad request" if a relative spec is used in `assignReference` or `deleteReference`
  a. This would work too, but it might not solve the issue completely because instead of conflict, we would throw a bad-request error.

Regarding option 2: this was actually my first idea. It solves the issue at hand, and doesn't seem to have any concerning downsides. I gave up on it, thinking that conceptually, option 1 was better than 2... but now I am not sure anymore :-)

Let's try option 2 first and see how it goes. If it doesn't look good, then I will go with your suggestion (option 3).

Reply:
@dimas-b Seems like CI is happy with option 2!